api/grpcserver/globalstate_service.go

```diff
@@ -270,33 +270,53 @@ func (s GlobalStateService) AccountDataStream(in *pb.AccountDataStreamRequest, s
 	filterReceipt := in.Filter.AccountDataFlags&uint32(pb.AccountDataFlag_ACCOUNT_DATA_FLAG_TRANSACTION_RECEIPT) != 0
 
 	// Subscribe to the various streams
-	var accountCh, rewardsCh, receiptsCh <-chan interface{}
-
+	var (
+		accountCh       <-chan interface{}
+		rewardsCh       <-chan interface{}
+		receiptsCh      <-chan interface{}
+		accountBufFull  <-chan struct{}
+		rewardsBufFull  <-chan struct{}
+		receiptsBufFull <-chan struct{}
+	)
 	if filterAccount {
 		if accountSubscription := events.SubscribeAccount(); accountSubscription != nil {
 			defer closeSubscription(accountSubscription)
-			accountCh = consumeEvents(stream.Context(), accountSubscription.Out())
+
+			accountCh, accountBufFull = consumeEvents(stream.Context(), accountSubscription.Out())
 		}
 	}
 	if filterReward {
 		if rewardsSubscription := events.SubscribeRewards(); rewardsSubscription != nil {
 			defer closeSubscription(rewardsSubscription)
-			rewardsCh = consumeEvents(stream.Context(), rewardsSubscription.Out())
+			rewardsCh, rewardsBufFull = consumeEvents(stream.Context(), rewardsSubscription.Out())
 		}
 	}
 	if filterReceipt {
 		if receiptsSubscription := events.SubscribeReceipts(); receiptsSubscription != nil {
 			defer closeSubscription(receiptsSubscription)
-			receiptsCh = consumeEvents(stream.Context(), receiptsSubscription.Out())
+			receiptsCh, receiptsBufFull = consumeEvents(stream.Context(), receiptsSubscription.Out())
 		}
 	}
 
 	for {
 		select {
+		case <-accountBufFull:
+			log.Info("account buffer is full, shutting down")
+			return status.Error(codes.Canceled, errAccountBufferFull)
+		case <-rewardsBufFull:
+			log.Info("rewards buffer is full, shutting down")
+			return status.Error(codes.Canceled, errRewardsBufferFull)
+		case <-receiptsBufFull:
+			log.Info("receipts buffer is full, shutting down")
+			return status.Error(codes.Canceled, errReceiptsBufferFull)
 		case updatedAccountEvent, ok := <-accountCh:
 			if !ok {
-				log.Info("account stream buffer is full, shutting down")
-				return status.Errorf(codes.Canceled, "stream buffer is full")
+				// we could handle this more gracefully, by no longer listening
```

Comment:
i think this comment is pretty much a terrible idea, can we remove it everywhere please?
if one subscription gets terminated, we never should continue sending events from other subscriptions

also, if we are not going to close accountCh (and other channels) we don't need to handle `!ok` case here. otherwise you still need to return same error here

Reply:
sure, this comment originally existed, so it appeared again after I reverted a commit